java/ql/lib/semmle/code/java/dataflow/internal/TaintTrackingUtil.qll

```diff
@@ -523,6 +525,9 @@ private class FormatterCallable extends TaintPreservingCallable {
   }
 }
 
+/** Holds if taint may flow from the operand of a bitwise expression to its result. */
+private predicate bitwiseStep(Expr src, BitwiseExpr sink) { sink.(BinaryExpr).getAnOperand() = src }
```

Comment:
Isn't this step slightly odd to include in general?

Reply:
I was surprised by taint not being tracked in cases like the following:

```java
int flag = INTERESTING_FAG || UNINTERESTING_FLAG;
sink(flag);
```

So that's why I added it as a general step (I thought other people would expect it to work out-of-the-box). But I admit its applicability is limited, so if you think it doesn't add that much value generally, I can make it a specific step just for the query.

Reply:
It makes sense when specifically tracking flags, but I don't think it makes much sense otherwise. Global data flow on int-sized data from remote flow sources can already travel quite far, and I don't think we'll want to extend that with bitwise operations in general.

Reply:
Oh, now that I changed it back I remembered why I did it like this. Actually, it's the sanitizer which needs the step (because it uses `TaintTracking::localExprTaint`), not the `TaintTracking::Configuration` itself:

```codeql
private class IntentFlagsOrDataChangedSanitizer extends IntentUriPermissionManipulationSanitizer {
  IntentFlagsOrDataChangedSanitizer() {
    exists(MethodAccess ma, Method m |
      // ...
      m.hasName("removeFlags") and
      TaintTracking::localExprTaint(any(GrantReadUriPermissionFlag f).getAnAccess(),
        ma.getArgument(0)) and
      TaintTracking::localExprTaint(any(GrantWriteUriPermissionFlag f).getAnAccess(),
        ma.getArgument(0))
      or
      m.hasName("setFlags") and
      not TaintTracking::localExprTaint(any(GrantUriPermissionFlag f).getAnAccess(),
        ma.getArgument(0))
      // ...
    )
  }
}
```

Should I write a configuration specifically for tracking flag taint?

Reply:
Local taint tracking is just a simple transitive closure of a step relation (jump-to-def on `localExprTaint` to have a look). So I'd suggest something like:
```
private predicate flagTaintStep(DataFlow::Node n1, DataFlow::Node n2) {
  TaintTracking::localTaintStep(n1, n2) or
  // insert bitwise step here
}
```
and then use `flagTaintStep*` instead of `TaintTracking::localExprTaint`.

Reply:
That's elegant! Thanks for spoon-feeding the solution, this is a trick I'm going to remember well.

Done in 71ba9d6. Note that I used a slightly different name (named after what the predicate _does_, not after how I intend to use it). Also, since I had started adding additional taint steps to the query, I left the abstract step class because it won't hurt to have it available in the future.